csharp/src/Apache.Arrow.Adbc/Tracing/TracingConnectionImpl.cs

```diff
@@ -0,0 +1,170 @@
+        private static readonly ConcurrentDictionary<string, ActivityListener> s_listeners = new();
+        private readonly ConcurrentQueue<Activity> _activityQueue = new();
+        private readonly IReadOnlyDictionary<string, string>? _options;
+        private DirectoryInfo? _traceDirectory;
+
+        internal TracingConnectionImpl(IReadOnlyDictionary<string, string>? options = default)
+        {
+            _options = options ?? new Dictionary<string, string>();
+            EnsureTracing();
+        }
+
+        internal TracingConnectionImpl(bool isTracingEnabled, string traceLocation, int traceMaxFileSizeKb, int traceMaxFiles)
+        {
+            var options = new Dictionary<string, string>();
+            options[TracingOptions.Connection.Trace] = isTracingEnabled.ToString();
+            options[TracingOptions.Connection.TraceLocation] = traceLocation;
+            options[TracingOptions.Connection.TraceFileMaxSizeKb] = traceMaxFileSizeKb.ToString();
+            options[TracingOptions.Connection.TraceFileMaxFiles] = traceMaxFiles.ToString();
+            _options = options;
+            EnsureTracing();
+        }
+
+        public ActivitySource? ActivitySource { get; private set; }
+
+        internal static string GetProductVersion()
+        {
+            FileVersionInfo fileVersionInfo = FileVersionInfo.GetVersionInfo(Assembly.GetExecutingAssembly().Location);
+            return fileVersionInfo.ProductVersion ?? ProductVersionDefault;
+        }
+
+        private void EnsureTracing()
+        {
+            if (true || _options.TryGetValue(TracingOptions.Connection.Trace, out string? traceOption) && bool.TryParse(traceOption, out bool traceEnabled))
+            {
+
+                // TODO: Handle exceptions
+                if (_options?.TryGetValue(TracingOptions.Connection.TraceLocation, out string? traceLocation) != true || !Directory.Exists(traceLocation))
+                {
+                    string? traceLocationDefault = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+                    _traceDirectory = new DirectoryInfo(traceLocationDefault);
+                }
+                else
+                {
+                    // TODO: If not exist, try to create
+                    _traceDirectory = new DirectoryInfo(traceLocation);
+                }
+                // TODO: Check if folder is writable
+
+
+                // TODO: Determine the best handling of listener lifetimes.
+                // Key of listeners collection should be ouput file location
+                ActivityListener listener = s_listeners.GetOrAdd(s_activitySourceName + "." + _traceDirectory.FullName, (_) => new()
+                {
+                    ShouldListenTo = (source) => source.Name == s_activitySourceName,
+                    Sample = (ref ActivityCreationOptions<ActivityContext> options) => ActivitySamplingResult.AllDataAndRecorded,
+                    ActivityStarted = OnActivityStarted,
```

Comment:
Note: implementors of the ActivityListener need to provide delegates to handle the ActivityStarted and ActivityStopped events.